src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -0,0 +1,253 @@
+        // Object backendRolesObject = ObjectUtils.firstNonNull(claims.get("ebr"), claims.get("dbr"));
+        if (!claims.containsKey("dbr")) {
+            return null;
+        }
+
+        Object backendRolesObject = claims.get("dbr");
+        String[] backendRoles;
+
+        if (backendRolesObject == null) {
+            log.warn("This is a malformed On-behalf-of Token");
+            backendRoles = new String[0];
+        } else {
+            final String backendRolesClaim = backendRolesObject.toString();
+
+            // Extracting roles based on the compatibility mode
+            String decryptedBackendRoles = backendRolesClaim;
+            backendRoles = Arrays.stream(decryptedBackendRoles.split(",")).map(String::trim).toArray(String[]::new);
+        }
+
+        return backendRoles;
+    }
+
+    @Override
+    @SuppressWarnings("removal")
+    public AuthCredentials extractCredentials(RestRequest request, ThreadContext context) throws OpenSearchSecurityException {
+        final SecurityManager sm = System.getSecurityManager();
+
+        if (sm != null) {
+            sm.checkPermission(new SpecialPermission());
+        }
+
+        AuthCredentials creds = AccessController.doPrivileged(new PrivilegedAction<AuthCredentials>() {
+            @Override
+            public AuthCredentials run() {
+                return extractCredentials0(request);
+            }
+        });
+
+        return creds;
+    }
+
+    private AuthCredentials extractCredentials0(final RestRequest request) {
+        if (!oboEnabled) {
+            log.error("On-behalf-of authentication has been disabled");
+            return null;
+        }
+
+        if (jwtParser == null) {
+            log.error("Missing Signing Key. JWT authentication will not work");
+            return null;
+        }
+
+        String jwtToken = request.header(HttpHeaders.AUTHORIZATION);
+
+        if (jwtToken == null || jwtToken.length() == 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("No JWT token found in '{}' header", HttpHeaders.AUTHORIZATION);
+            }
+            return null;
+        }
+
+        if (!BEARER.matcher(jwtToken).matches()) {
+            jwtToken = null;
+        }
+
+        if (jwtToken != null && Pattern.compile(BEARER_PREFIX).matcher(jwtToken.toLowerCase()).find()) {
+            jwtToken = jwtToken.substring(jwtToken.toLowerCase().indexOf(BEARER_PREFIX) + BEARER_PREFIX.length());
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("No Bearer scheme found in header");
+            }
+        }
+
+        if (jwtToken == null) {
+            return null;
+        }
+
+        try {
+            Matcher matcher = PATTERN_PATH_PREFIX.matcher(request.path());
+            final String suffix = matcher.matches() ? matcher.group(2) : null;
+            if (request.method() == RestRequest.Method.POST && ON_BEHALF_OF_SUFFIX.equals(suffix)
```

Comment:
It may help with readability to encapsulate this logic in a method like `OnBehalfOfAuthenticator.isAllowedRequest(request)`

Reply:
I have updated the code.